### ISSUE TITLE
Unicode conversion mismatch with SAP

### ISSUE THREAD

**Issue by sbcgua** [abaplint/transpiler#1857]
A very narrow case. Well isolated here: https://github.com/sbcgua/ajson5/commit/4e1b5a47074289b33537fd037c8d9d68094f4b3f

Fails in SAP before, works after. Both work in transpiler.

Here it is a bit more complex to explain. In fact the is this unit test that didn't pass: https://github.com/sbcgua/ajson5/commit/4e1b5a47074289b33537fd037c8d9d68094f4b3f#diff-6c0798cdb693e060efdd6c53e586938a5c8b95cee8e846ea35651775ccf12230R735-R746

Point is that `"user \u00e9 \uD834\uDD1E"` contains lower case Hex number. And for some reason SAP (731) does not support that and converts it to `00` not `E9` char (can be seen in the debugger). And no error is raise, just silent misintepretation. So the solution `translate lv_hex to upper case` in the commit helped. But obviously a mismatch between the SAP and the transpiler

**Comment by mbtools** [comment]
fyi, the ABAP behaviour is documented (see Byte-Like Target Fields):

https://help.sap.com/doc/abapdocu_latest_index_htm/latest/en-US/ABENCONVERSION_TYPE_C.html

**Comment by oisee** [comment]
Measured the rule on a 7.5x system: `x = c` / `x = string` takes the leading hexadecimal digits only (upper case `0-9A-F`), pads an odd count with `0`, then cuts or pads with `00` to the field's length. The runtime keeps everything as long as the *first* character is a hex digit, so the field ends up holding characters that are not hex at all.

Reproducer:

```abap
DATA lv_hex TYPE x LENGTH 4.
DATA lv_c6 TYPE c LENGTH 6.
DATA lv_str TYPE string.
lv_str = 'ABG1'.  lv_hex = lv_str.  WRITE / lv_hex.
lv_str = 'AB CD'. lv_hex = lv_str.  WRITE / lv_hex.
lv_str = '0a1B'.  lv_hex = lv_str.  WRITE / lv_hex.
lv_str = '12ab'.  lv_hex = lv_str.  WRITE / lv_hex.
lv_c6 = 'AB'.     lv_hex = lv_c6.   WRITE / lv_hex.
```

| input | on a 7.5x system | here (runtime 2.13.x, main today) |
| --- | --- | --- |
| `ABG1` | `AB000000` | `ABG10000` |
| `AB CD` | `AB000000` | `AB CD000` |
| `0a1B` | `00000000` | `0a1B0000` |
| `12ab` | `12000000` | `12ab0000` |
| c(6) `AB` | `AB000000` | `AB    00` |

The system column was measured with an ABAP Unit test on an on-premise 7.5x sandbox. Into an `xstring`, the runtime already takes the same prefix, so only `x` is affected. The branch is in `Hex.set` (`packages/runtime/src/types/hex.ts`), which keeps `v` whole and only clears it when the first character is not a hex digit.

A PR with the fix and a test follows, against this issue.
